### ISSUE TITLE
Github workflow downloaded files will not behave as expected

### ISSUE THREAD

**Issue by arthurp** [KatanaGraph/katana#129]
The github `actions/cache` behavior is not what you might expect: It does not update the cache after a run that started with a cache hit. Instead it only updates the cache after runs that started with a cache miss. This means that if a downloaded data file is updated (https://github.com/KatanaGraph/katana/blob/master/.github/workflows/cpp.yaml#L163-L176) the cache will not be rewritten so the file will be redownloaded every CI run there after (as long as the cache survives). Similarly, ccache caches will not be rewritten, so they will go stale rapidly as source files changes (since the source files are not part of the CI level cache hash).

Basically, we are caching data, but the cached data is not being updated by each run, so it is going stale even though it is still kept and loaded into every CI run.

To see evidence of this behavior look at: https://github.com/KatanaGraph/katana/runs/825080264?check_suite_focus=true and open the "Post Cache graph inputs" and "Post Cache ccache objects" steps. They report "Cache hit occurred on the primary key generic-cache-graph, not saving cache."

To be clear, none of this creates incorrect results. Instead, caching is just doing nothing and wasting a few resources.

**Comment by arthurp** [comment]
I suspect the solution to this is to create our own variant of actions/cache that writes to the cache after every run. This would give use the behavior we expected and need. For downloaded files a download-and-cache action would be even better, that understands the whole HTTP file date and e-tag system and can use it to download new copies exactly when needed.

**Comment by ddn0** [comment]
Thanks for looking into this :D I confess I assumed the behavior of the cache without actually verifying it :blush: 

For ccache, we can probably just drop caching ccache entirely. I don't think the win is that much from ccache is that much (especially since we can't share the ccache between multiple workflows run from the same source directory).

Looks like other people have noticed this deficiency with action/cache: https://github.com/actions/cache/pull/353 I propose we keep our current downloading code and wait for the "force update" option to be available.

**Comment by arthurp** [comment]
I actually suspect we could gain a lot from ccache if it was working (it's unlikely it has been doing anything at all for a while). I may do a build time study with ccache to see how much time gets saved during a build with successive revisions. Not important, but I'm really curious now. Maybe I'll write a script to collect data on this after work tonight.

BTW, we may also be running up against the 5GB caching limit.

**Comment by ddn0** [comment]
> Not important, but I'm really curious now. Maybe I'll write a script to collect data on this after work tonight.

Cool. Go for it.

Fast CI = Happy Developers !

> BTW, we may also be running up against the 5GB caching limit.

:+1:

**Comment by ddn0** [comment]
>> This means that if a downloaded data file is updated (https://github.com/KatanaGraph/katana/blob/master/.github/workflows/cpp.yaml#L163-L176) the cache will not be rewritten so the file will be redownloaded every CI run there after (as long as the cache survives).
> Looks like other people have noticed this deficiency with action/cache: actions/cache#353 I propose we keep our current downloading code and wait for the "force update" option to be available.

Actually thinking about it more, making sure that CI changes have a matching state change in the repo seems pretty reasonable. This goes back to the idea of us controlling dependency changes rather than having them silently occur. My new proposal is just a hashfile under `.github/` that we use in `hashFile` expressions and that we bump whenever we want to force a new download. Technically we control the graph input URL so we could also just give it a new name which boils down to the same thing. But since we can't change the miniconda URL, this will consistently handle both cases.

(I'm working on adding more testing data to enable testing the distributed routines so this problem happens to be top of my mind for me.)

This is probably a different use case from ccache.

**Comment by arthurp** [comment]
I'm a bit worried about this approach because the cache could be dropped due to cache usage pressure (5GB limit) causing the file to get updated even if we didn't want it to be. So this doesn't actually give us any control over the dependencies.

**Comment by ddn0** [comment]
> I'm a bit worried about this approach because the cache could be dropped due to cache usage pressure (5GB limit) causing the file to get updated even if we didn't want it to be. So this doesn't actually give us any control over the dependencies.

Fair enough; I'll just version the names of graph inputs and CI inputs that we control; then we just `hashFile` the yaml workflow files themselves. What you do for miniconda is still open. (We could, for instance, host and name that ourselves too.)

**Comment by arthurp** [comment]
@insertinterestingnamehere, is there a way to download the miniconda installer with a specific version or release date?

**Comment by insertinterestingnamehere** [comment]
Yah, just use a versioned link from https://repo.anaconda.com/miniconda/ instead of the latest one.

**Comment by insertinterestingnamehere** [comment]
For example: https://repo.anaconda.com/miniconda/Miniconda3-py38_4.8.3-Linux-x86_64.sh

**Comment by arthurp** [comment]
Of course! I didn't even think about how there was "latest" in the URL.

@ddn0 we can do the same thing you were talking about for Miniconda. Feel free to do that when you update other stuff. Or I can do it if you prefer.

**Comment by ddn0** [comment]
Cool! I'll add this when I add the new downloads.

**Comment by ddn0** [comment]
Updated this issue to reflect just fixing the downloads.

Feel free to file a separate issue for ccache if we have something actionable. Otherwise, I'll plan to just remove ccache in the CI for now.
